File: packages/domain/src/ledger.ts
import type { Result } from './shared/result';
import { err, ok } from './shared/result';
import type { IsoDate } from './shared/time';
import { addDaysIso, isSameOrBefore, msOf } from './shared/time';

export const LEDGER_ENTRY_TYPES = [
  'TOP_UP',
  'VISIT_DEDUCTION',
  'REFUND',
  'BONUS',
  'PROMO',
  'EXPIRATION',
  'ADJUSTMENT',
  'REVERSAL',
] as const;
export type LedgerEntryType = (typeof LEDGER_ENTRY_TYPES)[number];

export type LedgerSourceType = 'PAYMENT' | 'BOOKING' | 'ACCESS' | 'ADMIN' | 'SYSTEM';

/**
 * A finalized ledger entry is immutable. Corrections are REVERSAL entries
 * referencing the original — never edits, never deletes.
 */
export interface CreditLedgerEntry {
  id: string;
  memberId: string;
  type: LedgerEntryType;
  /** Signed integer credits. */
  amount: number;
  description: string;
  sourceType: LedgerSourceType | null;
  sourceId: string | null;
  /** Set on REVERSAL entries: the entry being reversed. */
  reversesEntryId: string | null;
  actorId: string | null;
  reason: string | null;
  createdAt: IsoDate;
}

/** A top-up lot tracks a batch of credits and when they expire (FIFO consumption). */
export interface TopUpLot {
  id: string;
  memberId: string;
  ledgerEntryId: string;
  /** The purchased package (null for bonus/adjustment credits). */
  packageId: string | null;
  credits: number;
  expiresAt: IsoDate;
  createdAt: IsoDate;
}

/** THE wallet rule: balance is always derived, never stored. */
export const computeBalance = (entries: readonly CreditLedgerEntry[]): number =>
  entries.reduce((sum, e) => sum + e.amount, 0);

export interface LotRemainder {
  lot: TopUpLot;
  remaining: number;
}

/**
 * Allocates consumption across lots FIFO (earliest expiry first) and returns
 * what is left in each lot. EXPIRATION entries are pinned to their lot via sourceId.
 */
export function computeLotRemainders(
  lots: readonly TopUpLot[],
  entries: readonly CreditLedgerEntry[],
): LotRemainder[] {
  const sorted = [...lots].sort((a, b) => msOf(a.expiresAt) - msOf(b.expiresAt));

  // Consumption pool: everything negative except EXPIRATION (those are lot-specific),
  // minus positive restorations (REFUND / positive REVERSAL give credits back).
  let consumed = 0;
  for (const e of entries) {
    if (e.type === 'EXPIRATION') continue;
    if (e.amount < 0) consumed += -e.amount;
    else if ((e.type === 'REFUND' || e.type === 'REVERSAL') && e.amount > 0) consumed -= e.amount;
  }
  consumed = Math.max(0, consumed);

  const expiredByLot = new Map<string, number>();
  for (const e of entries) {
    if (e.type === 'EXPIRATION' && e.sourceId) {
      expiredByLot.set(e.sourceId, (expiredByLot.get(e.sourceId) ?? 0) + -e.amount);
    }
  }

  return sorted.map((lot) => {
    const afterExpiry = lot.credits - (expiredByLot.get(lot.id) ?? 0);
    const take = Math.min(afterExpiry, consumed);
    consumed -= take;
    return { lot, remaining: afterExpiry - take };
  });
}

export interface DraftExpirationEntry {
  memberId: string;
  type: 'EXPIRATION';
  amount: number;
  description: string;
  sourceType: 'SYSTEM';
  sourceId: string;
}

/** Emits the EXPIRATION drafts for lots past their expiry that still hold credits. */
export function deriveExpirationEntries(
  lots: readonly TopUpLot[],
  entries: readonly CreditLedgerEntry[],
  now: IsoDate,
): DraftExpirationEntry[] {
  return computeLotRemainders(lots, entries)
    .filter(({ lot, remaining }) => remaining > 0 && isSameOrBefore(lot.expiresAt, now))
    .map(({ lot, remaining }) => ({
      memberId: lot.memberId,
      type: 'EXPIRATION' as const,
      amount: -remaining,
      description: `Credits expired (lot ${lot.id})`,
      sourceType: 'SYSTEM' as const,
      sourceId: lot.id,
    }));
}

/** Credits that will expire within `withinDays` (excluding already-expired lots). */
export function computeExpiringCredits(
  lots: readonly TopUpLot[],
  entries: readonly CreditLedgerEntry[],
  now: IsoDate,
  withinDays: number,
): number {
  const horizon = addDaysIso(now, withinDays);
  return computeLotRemainders(lots, entries)
    .filter(
      ({ lot, remaining }) =>
        remaining > 0 && msOf(lot.expiresAt) > msOf(now) && isSameOrBefore(lot.expiresAt, horizon),
    )
    .reduce((sum, { remaining }) => sum + remaining, 0);
}

export type ReversalProblem =
  | { type: 'CANNOT_REVERSE_REVERSAL' }
  | { type: 'ALREADY_REVERSED' };

export interface DraftReversalEntry {
  memberId: string;
  type: 'REVERSAL';
  amount: number;
  description: string;
  sourceType: 'ADMIN';
  sourceId: string | null;
  reversesEntryId: string;
  actorId: string;
  reason: string;
}

/** Builds the compensating entry for `original`. Caller must confirm no prior reversal exists. */
export function buildReversalEntry(
  original: CreditLedgerEntry,
  args: { actorId: string; reason: string; alreadyReversed: boolean },
): Result<DraftReversalEntry, ReversalProblem> {
  if (original.type === 'REVERSAL') return err({ type: 'CANNOT_REVERSE_REVERSAL' });
  if (args.alreadyReversed) return err({ type: 'ALREADY_REVERSED' });
  return ok({
    memberId: original.memberId,
    type: 'REVERSAL',
    amount: -original.amount,
    description: `Reversal of ${original.type} (${original.id})`,
    sourceType: 'ADMIN',
    sourceId: original.sourceId,
    reversesEntryId: original.id,
    actorId: args.actorId,
    reason: args.reason,
  });
}

File: internal/domain/ledger.go
package domain

import (
	"errors"
	"fmt"
	"sort"
	"time"
)

// LedgerEntryType classifies a credit movement.
type LedgerEntryType string

const (
	LedgerTopUp          LedgerEntryType = "TOP_UP"
	LedgerVisitDeduction LedgerEntryType = "VISIT_DEDUCTION"
	LedgerRefund         LedgerEntryType = "REFUND"
	LedgerBonus          LedgerEntryType = "BONUS"
	LedgerPromo          LedgerEntryType = "PROMO"
	LedgerExpiration     LedgerEntryType = "EXPIRATION"
	LedgerAdjustment     LedgerEntryType = "ADJUSTMENT"
	LedgerReversal       LedgerEntryType = "REVERSAL"
)

// LedgerSourceType names the subsystem that caused an entry, so any credit
// movement can be traced back to the payment, booking or scan behind it.
type LedgerSourceType string

const (
	SourcePayment LedgerSourceType = "PAYMENT"
	SourceBooking LedgerSourceType = "BOOKING"
	SourceAccess  LedgerSourceType = "ACCESS"
	SourceAdmin   LedgerSourceType = "ADMIN"
	SourceSystem  LedgerSourceType = "SYSTEM"
)

// CreditLedgerEntry is an immutable credit movement. Rows are never updated or
// deleted: a mistake is corrected by writing a REVERSAL that points at the
// original, so the history stays auditable.
type CreditLedgerEntry struct {
	ID       string          `json:"id"`
	MemberID string          `json:"memberId"`
	Type     LedgerEntryType `json:"type"`
	// Amount is signed: positive adds credits, negative consumes them.
	Amount          int               `json:"amount"`
	Description     string            `json:"description"`
	SourceType      *LedgerSourceType `json:"sourceType"`
	SourceID        *string           `json:"sourceId"`
	ReversesEntryID *string           `json:"reversesEntryId"`
	ActorID         *string           `json:"actorId"`
	Reason          *string           `json:"reason"`
	CreatedAt       time.Time         `json:"createdAt"`
}

// TopUpLot tracks one batch of purchased credits and when it expires. Lots
// exist so expiry can be FIFO: the credits that expire soonest are spent first.
type TopUpLot struct {
	ID            string `json:"id"`
	MemberID      string `json:"memberId"`
	LedgerEntryID string `json:"ledgerEntryId"`
	// PackageID is nil for credits that were not purchased (bonus, adjustment).
	PackageID *string   `json:"packageId"`
	Credits   int       `json:"credits"`
	ExpiresAt time.Time `json:"expiresAt"`
	CreatedAt time.Time `json:"createdAt"`
}

// ComputeBalance is THE wallet rule: a balance is the sum of the ledger, never
// a stored column that could drift out of sync with its history.
func ComputeBalance(entries []CreditLedgerEntry) int {
	total := 0
	for _, e := range entries {
		total += e.Amount
	}
	return total
}

// LotRemainder is how much of a lot is still unspent.
type LotRemainder struct {
	Lot       TopUpLot
	Remaining int
}

// ComputeLotRemainders allocates all consumption across lots in expiry order.
//
// Consumption is pooled rather than matched entry-by-entry: every negative
// entry except EXPIRATION spends from the pool, and credits handed back
// (REFUND, or a REVERSAL that adds) reduce it again. EXPIRATION is excluded
// because those entries are pinned to a specific lot through SourceID.
func ComputeLotRemainders(lots []TopUpLot, entries []CreditLedgerEntry) []LotRemainder {
	sorted := make([]TopUpLot, len(lots))
	copy(sorted, lots)
	sort.SliceStable(sorted, func(i, j int) bool {
		return sorted[i].ExpiresAt.Before(sorted[j].ExpiresAt)
	})

	consumed := 0
	expiredByLot := map[string]int{}
	for _, e := range entries {
		if e.Type == LedgerExpiration {
			if e.SourceID != nil {
				expiredByLot[*e.SourceID] += -e.Amount
			}
			continue
		}
		switch {
		case e.Amount < 0:
			consumed += -e.Amount
		case (e.Type == LedgerRefund || e.Type == LedgerReversal) && e.Amount > 0:
			consumed -= e.Amount
		}
	}
	if consumed < 0 {
		consumed = 0
	}

	out := make([]LotRemainder, 0, len(sorted))
	for _, lot := range sorted {
		afterExpiry := lot.Credits - expiredByLot[lot.ID]
		take := afterExpiry
		if consumed < take {
			take = consumed
		}
		if take < 0 {
			take = 0
		}
		consumed -= take
		out = append(out, LotRemainder{Lot: lot, Remaining: afterExpiry - take})
	}
	return out
}

// DraftExpirationEntry is an EXPIRATION the sweep should write.
type DraftExpirationEntry struct {
	MemberID    string
	Amount      int
	Description string
	LotID       string
}

// DeriveExpirationEntries returns the entries needed to zero out lots that are
// past their expiry but still hold credits. Running it twice is safe: the
// second pass sees the credits already gone and emits nothing.
func DeriveExpirationEntries(lots []TopUpLot, entries []CreditLedgerEntry, now time.Time) []DraftExpirationEntry {
	var drafts []DraftExpirationEntry
	for _, r := range ComputeLotRemainders(lots, entries) {
		if r.Remaining > 0 && !r.Lot.ExpiresAt.After(now) {
			drafts = append(drafts, DraftExpirationEntry{
				MemberID:    r.Lot.MemberID,
				Amount:      -r.Remaining,
				Description: fmt.Sprintf("Credits expired (lot %s)", r.Lot.ID),
				LotID:       r.Lot.ID,
			})
		}
	}
	return drafts
}

// ComputeExpiringCredits counts credits expiring within the horizon, excluding
// lots that already expired.
func ComputeExpiringCredits(lots []TopUpLot, entries []CreditLedgerEntry, now time.Time, withinDays int) int {
	horizon := now.AddDate(0, 0, withinDays)
	total := 0
	for _, r := range ComputeLotRemainders(lots, entries) {
		if r.Remaining > 0 && r.Lot.ExpiresAt.After(now) && !r.Lot.ExpiresAt.After(horizon) {
			total += r.Remaining
		}
	}
	return total
}

// Reversal errors. Both are business outcomes the API surfaces by code.
var (
	ErrCannotReverseReversal = errors.New("CANNOT_REVERSE_REVERSAL")
	ErrAlreadyReversed       = errors.New("ALREADY_REVERSED")
)

// DraftReversalEntry is the compensating entry for a mistaken movement.
type DraftReversalEntry struct {
	MemberID        string
	Amount          int
	Description     string
	SourceID        *string
	ReversesEntryID string
	ActorID         string
	Reason          string
}

// BuildReversal produces the entry that cancels out `original`. A reversal
// cannot itself be reversed, and an entry can only be reversed once.
func BuildReversal(original CreditLedgerEntry, actorID, reason string, alreadyReversed bool) (DraftReversalEntry, error) {
	if original.Type == LedgerReversal {
		return DraftReversalEntry{}, ErrCannotReverseReversal
	}
	if alreadyReversed {
		return DraftReversalEntry{}, ErrAlreadyReversed
	}
	return DraftReversalEntry{
		MemberID:        original.MemberID,
		Amount:          -original.Amount,
		Description:     fmt.Sprintf("Reversal of %s (%s)", original.Type, original.ID),
		SourceID:        original.SourceID,
		ReversesEntryID: original.ID,
		ActorID:         actorID,
		Reason:          reason,
	}, nil
}
